File: backend/Services/TaskService.cs
using Microsoft.EntityFrameworkCore;
using backend.Data;
using backend.Models;
using backend.Models.DTOs;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace backend.Services
{
    public class TaskService : ITaskService
    {
        private readonly AppDbContext _context;
        private readonly IProjectService _projectService; // Inject ProjectService for ownership check

        public TaskService(AppDbContext context, IProjectService projectService)
        {
            _context = context;
            _projectService = projectService;
        }

        // Helper to map ProjectTask entity to TaskDto
        private static TaskDto MapTaskToDto(ProjectTask task)
        {
            return new TaskDto
            {
                Id = task.Id,
                Title = task.Title,
                DueDate = task.DueDate,
                IsCompleted = task.IsCompleted,
                CreatedAt = task.CreatedAt,
                ProjectId = task.ProjectId
            };
        }

        public async Task<IEnumerable<TaskDto>> GetProjectTasksAsync(int projectId, int userId)
        {
            // First, verify the user owns the project
            if (!await _projectService.UserOwnsProjectAsync(projectId, userId))
            {
                 // Return empty list or throw an exception - returning empty is safer
                return Enumerable.Empty<TaskDto>();
            }

            return await _context.ProjectTasks
                .Where(t => t.ProjectId == projectId)
                .OrderBy(t => t.CreatedAt) // Order by creation time
                .Select(t => MapTaskToDto(t)) // Use mapping function
                .ToListAsync();
        }

        public async Task<TaskDto?> GetTaskByIdAsync(int taskId, int userId)
        {
            var task = await _context.ProjectTasks
                .Include(t => t.Project) // Include project to check ownership
                .FirstOrDefaultAsync(t => t.Id == taskId);

            // Check if task exists AND if the user owns the parent project
            if (task == null || task.Project == null || task.Project.UserId != userId)
            {
                return null;
            }

            return MapTaskToDto(task);
        }

        public async Task<TaskDto?> CreateTaskAsync(CreateTaskDto createDto, int userId)
        {
            // Verify the user owns the target project
            if (!await _projectService.UserOwnsProjectAsync(createDto.ProjectId, userId))
            {
                return null; // User doesn't own the project
            }

            var task = new ProjectTask
            {
                Title = createDto.Title,
                DueDate = createDto.DueDate,
                IsCompleted = false, // Default to not completed
                ProjectId = createDto.ProjectId,
                CreatedAt = DateTime.UtcNow
            };

            _context.ProjectTasks.Add(task);
            await _context.SaveChangesAsync();

            return MapTaskToDto(task);
        }

        public async Task<TaskDto?> UpdateTaskAsync(int taskId, UpdateTaskDto updateDto, int userId)
        {
            var task = await _context.ProjectTasks
                .Include(t => t.Project) // Include project for ownership check
                .FirstOrDefaultAsync(t => t.Id == taskId);

            // Check ownership
            if (task == null || task.Project == null || task.Project.UserId != userId)
            {
                return null; // Task not found or user doesn't own it
            }

            // Update properties
            task.Title = updateDto.Title;
            task.DueDate = updateDto.DueDate;
            task.IsCompleted = updateDto.IsCompleted;

            _context.ProjectTasks.Update(task);
            await _context.SaveChangesAsync();

            return MapTaskToDto(task);
        }

         public async Task<TaskDto?> ToggleTaskCompletionAsync(int taskId, int userId)
        {
             var task = await _context.ProjectTasks
                .Include(t => t.Project) // Include project for ownership check
                .FirstOrDefaultAsync(t => t.Id == taskId);

            // Check ownership
            if (task == null || task.Project == null || task.Project.UserId != userId)
            {
                return null; // Task not found or user doesn't own it
            }

            // Toggle completion status
            task.IsCompleted = !task.IsCompleted;

            _context.ProjectTasks.Update(task);
            await _context.SaveChangesAsync();

            return MapTaskToDto(task);
        }


        public async Task<bool> DeleteTaskAsync(int taskId, int userId)
        {
            var task = await _context.ProjectTasks
                .Include(t => t.Project) // Include project for ownership check
                .FirstOrDefaultAsync(t => t.Id == taskId);

            // Check ownership
            if (task == null || task.Project == null || task.Project.UserId != userId)
            {
                return false; // Task not found or user doesn't own it
            }

            _context.ProjectTasks.Remove(task);
            await _context.SaveChangesAsync();
            return true;
        }
    }
}

File: frontend/src/services/taskService.ts
import apiClient from './api';
import { Task, CreateTaskDto, UpdateTaskDto } from '../types/task.types';

export const taskService = {
  async getByProjectId(projectId: string): Promise<Task[]> {
    // Backend endpoint is /api/tasks/project/{projectId}
    const response = await apiClient.get<any[]>(`/tasks/project/${projectId}`);
    
    return response.data.map(t => ({
      id: String(t.id),
      title: t.title,
      dueDate: t.dueDate,
      isCompleted: t.isCompleted,
      projectId: String(t.projectId),
      createdAt: t.createdAt
    }));
  },

  async create(data: CreateTaskDto): Promise<Task> {
    // Convert projectId to number for backend
    const createData = {
      ...data,
      projectId: Number(data.projectId)
    };
    
    const response = await apiClient.post<any>('/tasks', createData);
    const task = response.data;
    
    return {
      id: String(task.id),
      title: task.title,
      dueDate: task.dueDate,
      isCompleted: task.isCompleted,
      projectId: String(task.projectId),
      createdAt: task.createdAt
    };
  },

  async update(id: string, data: UpdateTaskDto): Promise<Task> {
    const response = await apiClient.put<any>(`/tasks/${id}`, data);
    const task = response.data;
    
    return {
      id: String(task.id),
      title: task.title,
      dueDate: task.dueDate,
      isCompleted: task.isCompleted,
      projectId: String(task.projectId),
      createdAt: task.createdAt
    };
  },

  async toggleComplete(id: string): Promise<Task> {
    const response = await apiClient.patch<any>(`/tasks/${id}/toggle`);
    const task = response.data;
    
    return {
      id: String(task.id),
      title: task.title,
      dueDate: task.dueDate,
      isCompleted: task.isCompleted,
      projectId: String(task.projectId),
      createdAt: task.createdAt
    };
  },

  async delete(id: string): Promise<void> {
    await apiClient.delete(`/tasks/${id}`);
  }
};
